hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java

```diff
@@ -104,15 +105,15 @@ public static Schema addMetadataFields(Schema schema) {
     List<Schema.Field> parentFields = new ArrayList<>();
 
     Schema.Field commitTimeField =
-        new Schema.Field(HoodieRecord.COMMIT_TIME_METADATA_FIELD, METADATA_FIELD_SCHEMA, "", (Object) null);
+        new Schema.Field(HoodieRecord.COMMIT_TIME_METADATA_FIELD, METADATA_FIELD_SCHEMA, "", NullNode.getInstance());
     Schema.Field commitSeqnoField =
-        new Schema.Field(HoodieRecord.COMMIT_SEQNO_METADATA_FIELD, METADATA_FIELD_SCHEMA, "", (Object) null);
+        new Schema.Field(HoodieRecord.COMMIT_SEQNO_METADATA_FIELD, METADATA_FIELD_SCHEMA, "", NullNode.getInstance());
     Schema.Field recordKeyField =
-        new Schema.Field(HoodieRecord.RECORD_KEY_METADATA_FIELD, METADATA_FIELD_SCHEMA, "", (Object) null);
+        new Schema.Field(HoodieRecord.RECORD_KEY_METADATA_FIELD, METADATA_FIELD_SCHEMA, "", NullNode.getInstance());
     Schema.Field partitionPathField =
-        new Schema.Field(HoodieRecord.PARTITION_PATH_METADATA_FIELD, METADATA_FIELD_SCHEMA, "", (Object) null);
+        new Schema.Field(HoodieRecord.PARTITION_PATH_METADATA_FIELD, METADATA_FIELD_SCHEMA, "", NullNode.getInstance());
     Schema.Field fileNameField =
-        new Schema.Field(HoodieRecord.FILENAME_METADATA_FIELD, METADATA_FIELD_SCHEMA, "", (Object) null);
+        new Schema.Field(HoodieRecord.FILENAME_METADATA_FIELD, METADATA_FIELD_SCHEMA, "", NullNode.getInstance());
```

Comment:
By doing this change, we are effectively going back to deprecated constructors again. Can you please explain the purpose of doing this?

Reply:
it doesn't generate correct avro schema. This might not be the correct fix, but passing (Object) null results is the schema without default value.

Reply:
Ok.. I am still trying to understand what difference does it make if the schema is generated without defaultValue or if it is generated with defaultValue as null. Can you please provide more points to support this?

Reply:
Incorrect avro schema is a disaster waiting to happen. Current one says: it takes "null" or String, but if you try to read the message with reader schema that doesn't have those fields it will break as default will not be provided. Given that it is serialized in both log and parquet files it has to be correct, as if we add new field it will not be readable. Here is example exception when message was written without a field and  attempted to be read with field that doesn't have default:
`Caused by: org.apache.avro.AvroTypeException: Found Event, expecting Event, missing required field exist
	at org.apache.avro.io.ResolvingDecoder.doAction(ResolvingDecoder.java:292)
	at org.apache.avro.io.parsing.Parser.advance(Parser.java:88)
	at org.apache.avro.io.ResolvingDecoder.readFieldOrder(ResolvingDecoder.java:130)
	at org.apache.avro.generic.GenericDatumReader.readRecord(GenericDatumReader.java:215)
	at org.apache.avro.generic.GenericDatumReader.readWithoutConversion(GenericDatumReader.java:175)
	at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:153)
	at org.apache.avro.generic.GenericDatumReader.read(GenericDatumReader.java:145)`

Reply:
Ok.. IMO hoodie metadata fields will never be written as null. If you strongly feel about doing this change, can we think of some alternate solution rather than going back to deprecated constructors. Also wanted to check with you how are you generating schema like you mentioned before?

Reply:
Sure, there might be another way. I think there are changes with nulls in 1.9.0, but upgrading is not easy. Just to clarify, it is not for the write support, but for the read one. Consider situation:
We have avro with 4 metada fields written, and defaults are not set. Not an issue as it is written.
We add another metadata field (sourceOffset, or something), which also doesn't have default field and redeploy. When we try to read old record with new schema it will fail because old record doesn't have new field and there is no default for it. I hit this issue with spark-> avro transformation, have another PR to address

Reply:
On schema generation, we had to do some stuff... Our service emits protobufs to kafka, which we transform to avro using ProtoToAvro converter. Because those are service to service messages, they are not perfect and we had to do some bending to make sure schema can evolve. Some things: 
1) We made everything optional
2) Avro and Protobuf allows records to have no fields. But parquet doesn't, so, we inject marker field called exists pretty much everywhere. Originally added it only to records with no fields, but found out that when fields are added and exists is autoremoved, parquet-avro reader breaks as it expects every field in parquet schema to be present in avro. I have a fix to parquet-avro to skip those fields, will send a PR to show so we can discuss whether it is a good idea or not. 
3) Then we sometimes run sql transformation (same kafka stream produces multiple outputs). In this case, we infer schema from Spark (using NullTargetSchemaProvider). This schema must be correct as well which is not the case as spark omits defaults, so compaction breaks. Have a PR to adress.

Reply:
> parquet-avro reader breaks as it expects every field in parquet schema to be present in avro. I have a fix to parquet-avro to skip those fields, will send a PR to show so we can discuss whether it is a good idea or not

Yes, please. Would love to see how you are approaching that fix.

Reply:
Overloading of `null` to denote absence of a default value in avro, is confusing to say the least.. (atleast I learnt that now) But I am trying to grok the actual change in behavior.. 

This change seems to be orthogonal to the fix below.. Effectively, we are making the metadata fields nullable (as opposed to having no default values), with this change... While I agree, metadata fields won't have nulls.. Having this safety is better..